akka-actor/src/main/scala/akka/pattern/RetrySupport.scala

```diff
@@ -37,13 +36,60 @@ trait RetrySupport {
   def retry[T](attempt: () => Future[T], attempts: Int, delay: FiniteDuration)(
       implicit ec: ExecutionContext,
       scheduler: Scheduler): Future[T] = {
+    retry(attempt, attempts, _ => Some(delay))
+  }
+
+  /**
+   * Given a function from Unit to Future, returns an internally retrying Future.
+   * The first attempt will be made immediately, each subsequent attempt will be made after 'delay'.
+   * A scheduler (eg context.system.scheduler) must be provided to delay each retry.
+   * You could provide a function to generate the next delay duration after first attempt,
+   * this function should never return `null`, otherwise an [[IllegalArgumentException]] will be through.
+   * If attempts are exhausted the returned future is simply the result of invoking attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent
+   * tries and therefore must be thread safe (not touch unsafe mutable state).
+   *
+   * <b>Example usage:</b>
+   *
+   * //retry with back off
+   * {{{
+   * protected val sendAndReceive: HttpRequest => Future[HttpResponse]
+   * private val sendReceiveRetry: HttpRequest => Future[HttpResponse] = (req: HttpRequest) => retry[HttpResponse](
+   *   attempt = () => sendAndReceive(req),
+   *   attempts = 10,
+   *   delayFunction = attempted => Option(2.seconds * attempted)
+   *   scheduler = context.system.scheduler
+   * )
+   * }}}
+   */
+  def retry[T](attempt: () => Future[T], attempts: Int, delayFunction: Int => Option[FiniteDuration])(
+      implicit
+      ec: ExecutionContext,
+      scheduler: Scheduler): Future[T] = {
+    RetrySupport.retry(attempts, attempt, attempted = 0, delayFunction)
+  }
+}
+
+object RetrySupport extends RetrySupport {
+
+  private def retry[T](
+      maxAttempts: Int,
+      attempt: () => Future[T],
+      attempted: Int,
+      delayFunction: Int => Option[FiniteDuration])(implicit ec: ExecutionContext, scheduler: Scheduler): Future[T] = {
     try {
-      if (attempts > 0) {
+      if (maxAttempts - attempted > 0) {
         attempt().recoverWith {
           case NonFatal(_) =>
-            after(delay, scheduler) {
-              retry(attempt, attempts - 1, delay)
+            delayFunction(attempted) match {
```

Comment:
It seems that the `delayFunction` is returning an Option so users can decide if they want it or not. In that case, if it's optional, it will be better to have a overloaded method. One without a `delayFunction` and one with a delay function defined as `Int => FiniteDuration`.

Reply:
Moreover, if `delayFunction` returns a `null`, we will have a match error, not `IllegalArgumentException` as stated in the scaladoc. 

Maybe we should re-think the strategy here:

* if the `delayFunction` doesn't delay, we just call `attempt()` and Future complete with whatever attempt will return
* explicitly fail the Future with `IllegalArgumentException`. I find this better, at least we let the caller know they have misconfiure somethig. 
* eargerly call the function and throw `IllegalArgumentException` on user face. I like this better. :-)

Reply:
@renatocaval I have updated , please review.